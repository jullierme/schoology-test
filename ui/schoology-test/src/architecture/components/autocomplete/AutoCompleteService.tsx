import axios from 'axios';

export interface AutoCompleteService {
    fieldName: string;
    fieldValue: string;
    methodUrl: string;
    find(inputValue: string): Promise<any>;
    getUrl(inputValue: string): string;
}

export abstract class AutoCompleteServiceImpl implements AutoCompleteService {
    abstract fieldName: string;
    abstract fieldValue: string;
    methodUrl: string = '/country/find-by-name';

    getUrl(inputValue: string): string {
        return `${process.env.REACT_APP_API_URL}${
            this.methodUrl
        }/${encodeURIComponent(inputValue)}?size=7`;
    }

    find(inputValue: string): Promise<any> {
        inputValue = inputValue || '';

        return axios
            .get(this.getUrl(inputValue))
            .then(response => response.data)
            .then(response =>
                response.map((item: any) => {
                    item.value = item[this.fieldValue];
                    delete item[this.fieldValue];

                    item.label = item[this.fieldName];
                    delete item[this.fieldName];

                    //console.log(item);
                    return item;
                })
            );
    }
}
