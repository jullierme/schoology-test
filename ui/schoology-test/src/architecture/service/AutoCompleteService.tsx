export interface AutoCompleteService {
  fieldName: string;
  fieldValue: string;
  methodUrl: string;
}

export abstract class AutoCompleteServiceImpl
  implements AutoCompleteService {
  abstract fieldName: string;
  abstract fieldValue: string;
  methodUrl: string = '/country/find-by-name';
}
