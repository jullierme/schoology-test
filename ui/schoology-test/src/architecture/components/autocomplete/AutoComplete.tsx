import React from 'react';
import './AutoComplete.scss';
import { AutoCompleteService } from './AutoCompleteService';

import AsyncSelect from 'react-select/async';

interface AutocompleteProps {
    service: AutoCompleteService;
    value?: any;
    onChange?: any;
    style?: any;
}

const Autocomplete: React.FC<AutocompleteProps> = (
    props: AutocompleteProps
) => {
    const onChange = (item: any, actionMeta: any) => {
        if (
            actionMeta.action !== 'input-blur' &&
            actionMeta.action !== 'menu-close'
        ) {
            props.onChange(item);
        }
    };

    const loadOptions = (inputValue: string, callback: (data: any) => void) => {
        props.service.find(inputValue).then(callback);
    };

    return (
        <AsyncSelect
            id="asyncSelect"
            cacheOptions={false}
            defaultOptions={false}
            loadOptions={loadOptions}
            onChange={onChange}
        />
    );
};

export default Autocomplete;
