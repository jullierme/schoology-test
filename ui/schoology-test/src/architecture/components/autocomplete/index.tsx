import React from 'react';
import './style.scss';
import { AutoCompleteService } from '../../service/AutoCompleteService';

interface AutocompleteProps {
  label: string;
  service: AutoCompleteService;
  value: any;
  onChange?: any;
}

const Autocomplete: React.FC<AutocompleteProps> = (
  props: AutocompleteProps
) => {
  const onChangeAutoComplete = (event: any) => {
    props.onChange(event);
  };

  return (
    <>
      <label>{props.label}</label>
      <input
        type="text"
        value={props.value}
        onChange={onChangeAutoComplete}
      />
    </>
  );
};

export default Autocomplete;

/*
  /*function setModel(e: React.ChangeEvent<HTMLInputElement>) {
    const value = e.target.value;

    props.setService((prevState: any) => {
      const aux = { ...prevState };

      aux[props.name] = value;

      return aux;
    });
  }

Autocomplete.propTypes = {
  label: PropTypes.string.isRequired,
  model: PropTypes.object.isRequired,
  setModel: PropTypes.func.isRequired,
  name: PropTypes.string.isRequired,

  onChange: PropTypes.func,
  sizeList: PropTypes.number,
}*/

/*Autocomplete.defaultProps = {
  sizeList: config.get('autocomplete').defaultSizeList,
}*/
