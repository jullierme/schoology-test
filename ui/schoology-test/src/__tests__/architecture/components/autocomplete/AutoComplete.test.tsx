import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';

import Autocomplete from '../../../../architecture/components/autocomplete/AutoComplete';
import { AutoCompleteServiceImpl } from '../../../../architecture/components/autocomplete/AutoCompleteService';

import toJson from 'enzyme-to-json';

class TestAcService extends AutoCompleteServiceImpl {
    fieldName = 'name';
    fieldValue = 'abbreviation';
}

const service = new TestAcService();

it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Autocomplete service={service} />, div);

    ReactDOM.unmountComponentAtNode(div);
});

it('renders without crashing', () => {
    const wrapper = shallow(<Autocomplete service={service} />);

    expect(wrapper).toMatchSnapshot();

    console.log(toJson(wrapper));
});

