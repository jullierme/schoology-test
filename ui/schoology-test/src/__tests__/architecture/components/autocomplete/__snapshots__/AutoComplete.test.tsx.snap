// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<Async
  cacheOptions={false}
  defaultOptions={false}
  filterOption={null}
  id="asyncSelect"
  loadOptions={[Function]}
  onChange={[Function]}
/>
`;
